SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/Connection/SpatialEventTracer.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineMinimal.h"
+#include "GameFramework/Actor.h"
+#include "GenericPlatform/GenericPlatformAtomics.h"
+#include "Interop/Connection/SpatialConnectionManager.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "UObject/Class.h"
+#include <WorkerSDK/improbable/c_trace.h>
+#include <WorkerSDK/improbable/c_io.h>
+
+DEFINE_LOG_CATEGORY(LogSpatialEventTracer);
+
+using namespace SpatialGDK;
+using namespace worker::c;
+
+void SpatialEventTracer::TraceCallback(void* UserData, const Trace_Item* Item)
+{
+	SpatialEventTracer* EventTracer = static_cast<SpatialEventTracer*>(UserData);
+	checkf(EventTracer->bEnabled && EventTracer->Stream, TEXT("Tracer in an invalid state in TraceCallback")); 
+	if (EventTracer->BytesWrittenToStream < EventTracer->MaxFileSize)
+	{
+		uint32_t ItemSize = Trace_GetSerializedItemSize(Item); 
+		EventTracer->BytesWrittenToStream += ItemSize;
+		int Code = Trace_SerializeItemToStream(EventTracer->Stream, Item, ItemSize);
+		if (Code != 0)
+		{
+			UE_LOG(LogSpatialEventTracer, Error, TEXT("Failed to serialize to with error code %d (%s"), Code, Trace_GetLastError());
+		}
+	}
+}
+
+SpatialSpanIdActivator::SpatialSpanIdActivator(SpatialEventTracer* InEventTracer, const TOptional<Trace_SpanId>& InCurrentSpanId)
+	: CurrentSpanId(InCurrentSpanId)
+	, EventTracer(InEventTracer->GetWorkerEventTracer())
+{
+	if (InCurrentSpanId.IsSet())
+	{
+		Trace_EventTracer_SetActiveSpanId(EventTracer, CurrentSpanId.GetValue());
+	}
+}
+
+SpatialSpanIdActivator::~SpatialSpanIdActivator()
+{
+	if (CurrentSpanId.IsSet())
+	{
+		Trace_EventTracer_ClearActiveSpanId(EventTracer);
+	}
+}
+
+SpatialEventTracer::SpatialEventTracer(const FString& WorkerName)
+{
+	if (const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>())
+	{
+		if (Settings->bEventTracingEnabled)
+		{
+			MaxFileSize = Settings->MaxEventTracingFileSizeBytes;
+			Enable(WorkerName);
+		}
+	}
+}
+
+SpatialEventTracer::~SpatialEventTracer()
+{
+	if (IsEnabled())
+	{
+		Disable();
+	}
+}
+
+Trace_SpanId SpatialEventTracer::CreateNewSpanId()
+{
+	return Trace_EventTracer_AddSpan(EventTracer, nullptr, 0);
+}
+
+Trace_SpanId SpatialEventTracer::CreateNewSpanId(const TArray<Trace_SpanId>& Causes)
+{
+	return Trace_EventTracer_AddSpan(EventTracer, Causes.GetData(), Causes.Num());
+}
+
+TOptional<Trace_SpanId> SpatialEventTracer::TraceEvent(const FEventMessage& EventMessage, const UStruct* Struct, const worker::c::Trace_SpanId* Cause)
+{
+	if (!IsEnabled())
+	{
+		return {};
+	}
+
+	Trace_SpanId CurrentSpanId;
+	if (Cause)
+	{
+		CurrentSpanId = CreateNewSpanId({ *Cause });
+	}
+	else
+	{
+		CurrentSpanId = CreateNewSpanId();
```

Comment:
The way this whole `TraceEvent` function is implemented appears to enforce the constraint that each event should belong to a new span. There also doesn't appear to be a way of adding a span without adding an event or adding a span with more than 1 cause.

I guess my point is that this API appears to be significantly constrained compared to the underlying worker `EventTracer` API without an apparently good reason. It seems to me that it is probably useful to have a bunch of auxiliary functions which can do things like turn unreal objects into event data, but expose the semantics of the underlying APIs to the GDK codebase (whether that's by directly exposing the underlying event tracer object or by creating a dumb wrapper/proxy class).